paddle/phi/core/distributed/auto_parallel/inferspmd_utils.h

```diff
@@ -0,0 +1,171 @@
+      paddle::small_vector<Attribute, phi::kAttrSmallVectorSize> attrs)
+      : inputs_(std::move(inputs)), attrs_(std::move(attrs)) {}
+
+  void EmplaceBackInput(MetaTensor input);
+  void EmplaceBackAttr(Attribute attr);
+
+  const MetaTensor& InputAt(size_t idx) const;
+
+  template <typename AttrType>
+  AttrType AttrAt(size_t idx) const;
+
+  const Attribute& AttrAt(size_t idx) const;
+
+ private:
+  paddle::small_vector<MetaTensor, phi::kInputSmallVectorSize> inputs_;
+  paddle::small_vector<Attribute, phi::kAttrSmallVectorSize> attrs_;
+};
+
+using InferSpmdFn = SpmdInfo (*)(const InferSpmdContext&);
+
+#define PD_INFER_SPMD(...)                                    \
+  ::phi::distributed::InferSpmdFnImpl<decltype(&__VA_ARGS__), \
+                                      &__VA_ARGS__>::Call
+
+template <typename T>
+struct InferSpmdTypeTag {};
+
+template <typename Fn, Fn fn>
+struct InferSpmdFnImpl;
+
+template <typename Return, typename... Args, Return (*infer_spmd_fn)(Args...)>
+struct InferSpmdFnImpl<Return (*)(Args...), infer_spmd_fn> {
+  static SpmdInfo Call(const InferSpmdContext& ctx) {
+    return InferSpmdFnCallHelper<Args..., InferSpmdTypeTag<int>>::
+        template Call<0, 0>(ctx);
+  }
+
+ private:
+  template <typename... RemainingArgs>
+  struct InferSpmdFnCallHelper;
+
+  // TODO(chenweihang): support other input type later
+  template <typename... Tail>
+  struct InferSpmdFnCallHelper<const MetaTensor&, Tail...> {
+    template <int in_idx, int attr_idx, typename... PreviousArgs>
+    static SpmdInfo Call(const InferSpmdContext& ctx, PreviousArgs&... pargs) {
+      static_assert(attr_idx == 0,
+                    "InferSpmd's Input should appear before Attributes.");
+      const MetaTensor& arg = ctx.InputAt(in_idx);
+      return InferSpmdFnCallHelper<Tail...>::template Call<in_idx + 1,
+                                                           attr_idx>(
+          ctx, pargs..., arg);
+    }
+  };
+
+#define PD_SPECIALIZE_InferSpmdFnCallHelper_FOR_ATTRIBUTE(attr_type)      \
+  template <typename... Tail>                                             \
+  struct InferSpmdFnCallHelper<attr_type, Tail...> {                      \
+    template <int in_idx, int attr_idx, typename... PreviousArgs>         \
+    static SpmdInfo Call(const InferSpmdContext& ctx,                     \
+                         PreviousArgs&... pargs) {                        \
+      attr_type arg = ctx.AttrAt<attr_type>(attr_idx);                    \
+      return InferSpmdFnCallHelper<Tail...>::template Call<in_idx,        \
+                                                           attr_idx + 1>( \
+          ctx, pargs..., arg);                                            \
+    }                                                                     \
+  }
+
+  // TODO(chenweihang): support other attr type later by needed
+  PD_SPECIALIZE_InferSpmdFnCallHelper_FOR_ATTRIBUTE(bool);
```

Comment:
will this method cover complex attribute type like std::vector<int64_t>, std::vector<int>, std::vector<bool>

Reply:
yes, we can support these types later, now matmul cannot cover these types